### ISSUE TITLE
waf 2.0.0pre4: Problems with book example 11.4.2. A compiler producing source files with names unknown in advance

### ISSUE THREAD

**Issue by q-p** [waf-project/waf#1981]
In my build (creating C++ shared-libraries as well as Python extensions using those as well as generating doxygen documentation) I have something like the example given in the book ("scenarios_unknown"), where I am generating an unknown number of C++ files which should (additionally) be compiled into a shared library. This does not seem to work anymore when moving from waf 1.9 to 2.0.

The first problems that occurs is that the `tasks` attributes of the task-generator is missing:
```
Traceback (most recent call last):
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Scripting.py", line 122, in waf_entry_point
    run_commands()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Scripting.py", line 183, in run_commands
    ctx=run_command(cmd_name)
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Scripting.py", line 174, in run_command
    ctx.execute()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Scripting.py", line 368, in execute
    return execute_method(self)
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Build.py", line 93, in execute
    self.execute_build()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Build.py", line 100, in execute_build
    self.compile()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Build.py", line 167, in compile
    self.producer.start()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Runner.py", line 187, in start
    self.refill_task_list()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Runner.py", line 79, in refill_task_list
    self.get_out()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Runner.py", line 143, in get_out
    self.add_more_tasks(tsk)
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Runner.py", line 114, in add_more_tasks
    ready,waiting=self.prio_and_split(tsk.tasks)
AttributeError: 'src2c' object has no attribute 'tasks'
```

If that is "fixed" by setting an empty `tasks = []` attribute in the `src2c` class, then no error occurs, but the `more_tasks` generated are never actually executed.

Another potential problem, which I couldn't replicate with the book example but which occurs in my actual build is that after the "successful" build which is missing the execution of the `more_tasks` is that on a subsequent `waf build` I get the following error:
```
Traceback (most recent call last):
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Scripting.py", line 122, in waf_entry_point
    run_commands()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Scripting.py", line 183, in run_commands
    ctx=run_command(cmd_name)
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Scripting.py", line 174, in run_command
    ctx.execute()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Scripting.py", line 368, in execute
    return execute_method(self)
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Build.py", line 93, in execute
    self.execute_build()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Build.py", line 100, in execute_build
    self.compile()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Tools/errcheck.py", line 135, in check_compile
    ret=self.orig_compile()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Build.py", line 167, in compile
    self.producer.start()
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Runner.py", line 215, in start
    self.skip(tsk)
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Runner.py", line 152, in skip
    self.mark_finished(tsk)
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Runner.py", line 138, in mark_finished
    try_unfreeze(x)
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Runner.py", line 127, in try_unfreeze
    self.insert_with_prio(x)
  File ".waf-2.0.0-6ac063611e96930ef56af7eecd5d53cd/waflib/Runner.py", line 108, in insert_with_prio
    if self.outstanding and tsk.prio>=self.outstanding[0].prio:
AttributeError: 'cxx' object has no attribute 'prio'
```

What would be the updated way of achieving this scenario?

**Comment by ita1024** [comment]
Thanks for testing! While a new release is unavailable, one could try the commit above.

The examples will be included in regression test in the future.

**Comment by q-p** [comment]
Thanks, the commit fixes the above-mentioned issues. I'm still having problem with some files rebuilding for no obvious reason (but none of them are "generated" source-files), but that is probably a separate issue.

**Comment by ita1024** [comment]
I just noticed that the change was incomplete (https://waf.io/ci/). A fix was pushed as part of 3e1e6586e51c24a9ee9bce838eb0c1ee7609cba9.
